File: HOT100/12.栈/739.每日温度.cpp
/*
 * @lc app=leetcode.cn id=739 lang=cpp
 *
 * [739] 每日温度
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> stk;
        int n = temperatures.size();
        vector<int> res(n, 0);
        for (int i = 0; i < n; i ++ )
        {
            while (stk.size() && temperatures[i] > temperatures[stk.top()])
            {
                int pre = stk.top();
                stk.pop();
                res[pre] = i - pre;
            }
            stk.push(i);
        }
        return res;
    }
};
// @lc code=end

File: HOT100/12.栈/739.每日温度.py
#
# @lc app=leetcode.cn id=739 lang=python3
#
# [739] 每日温度
#
from mytools import *
# @lc code=start
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stk = []
        res = [0] * len(temperatures)
        stk.append(0)
        for i in range(1, len(temperatures)):
            while stk and temperatures[i] > temperatures[stk[-1]]:
                idx = stk.pop()
                res[idx] = i - idx
            stk.append(i)
        return res
# @lc code=end

